[9.0] report_py3o

faide opened the pull request:
This is the result of OCA days work in the "Open Document Reporting Workgroup"

We will add unit tests on this PR later on

The companion document is available here: https://docs.google.com/spreadsheets/d/1YgniNHjEWDHXZkhYWwasxmIklW734TL143euQ2Sjepc/

houzefa-abba committed:
Init - Print the odt template without modification (empty values)

houzefa-abba committed:
Overridable function to customize the values sent to the py3o renderer

houzefa-abba committed:
Update .hgignore

houzefa-abba committed:
Provide a lang object

houzefa-abba committed:
py3o.template now knows about the pyjon req

houzefa-abba committed:
Customization of report templates with a field to download/upload

houzefa-abba committed:
Hide the function field when the binary field is filled

houzefa-abba committed:
Add a function to format dates

houzefa-abba committed:
Version 1.0

houzefa-abba committed:
Added tag TAG_1.0 for changeset ae59b92cb6bd

houzefa-abba committed:
Declare the dependency on py3o.template - fix #3

Commit:
bumped version to 1.1 (stable)

Commit:
Added tag TAG_1.1 for changeset 392758bf61eb

Commit:
bumped version to 1.1.1 (dev)

anael-lorimier committed:
Now we call fusion server just to convert reports into the right format

anael-lorimier committed:
1.2

anael-lorimier committed:
Added tag TAG_1.2 for changeset b6c64ac6e399

anael-lorimier committed:
Remove deps

anael-lorimier committed:
1.2.1

anael-lorimier committed:
Added tag TAG_1.2.1 for changeset cd840bda3fa0

anael-lorimier committed:
Add missing files

anael-lorimier committed:
1.2.2

anael-lorimier committed:
Added tag TAG_1.2.2 for changeset 092018f8f6ca

anael-lorimier committed:
move py3o menu

anael-lorimier committed:
1.2.3

anael-lorimier committed:
Added tag TAG_1.2.3 for changeset 2203f0ccc2e2

anael-lorimier committed:
Add skipfusion flag to post

anael-lorimier committed:
1.2.4

anael-lorimier committed:
Added tag TAG_1.2.4 for changeset d6fa2a1477bc

anael-lorimier committed:
Add security rules

anael-lorimier committed:
1.2.5

anael-lorimier committed:
Added tag TAG_1.2.5 for changeset 067b603cc82e

anael-lorimier committed:
Call new methods from py3o.template to easily jsonify our datastruct

anael-lorimier committed:
Remove useless stuff

Commit:
1.3 production release

Commit:
Added tag TAG_1.3 for changeset f1771be362a2

houzefa-abba committed:
Version 1.3

--HG--
branch : 1.2

houzefa-abba committed:
Added tag TAG_1.3 for changeset ce6b34e12517

--HG--
branch : 1.2

houzefa-abba committed:
Merge the 1.2 branch

houzefa-abba committed:
Version 1.3.1

houzefa-abba committed:
Added tag TAG_1.3.1 for changeset a9d3f3747769

Commit:
Update and Add License (AGPL) and (XCG)

Commit:
Add README

anael-lorimier committed:
Odoo 8 fixes

--HG--
branch : odoo8

anael-lorimier committed:
Development of Py3o Parser to reproduce the behaviour of WebKit Report but using py3o engines

--HG--
branch : odoo8

anael-lorimier committed:
Py3o update for odoo 8

--HG--
branch : odoo8

anael-lorimier committed:
Fix openerp.py xml order

--HG--
branch : odoo8

faide committed:
we depend on report

--HG--
branch : odoo8

anael-lorimier committed:
Fix some pep8 and mistakes

--HG--
branch : odoo8

faide committed:
Real instructions to allow odoo to find the py3o.template module

--HG--
branch : odoo8

faide committed:
Just add the CSV to the data section

--HG--
branch : odoo8

faide committed:
pep ftw

--HG--
branch : odoo8

faide committed:
Output format is MANDATORY... if you don't give it will fail horribly (ie: fusion running at 100% cpu without traceback or info..) This is clearly a bug in fusion, but better fix it here too

--HG--
branch : odoo8

faide committed:
better docstring

--HG--
branch : odoo8

faide committed:
Adding a fallback system so that you own implementer module can define its default template

--HG--
branch : odoo8

faide committed:
removed unused file

--HG--
branch : odoo8

faide committed:
Now support latest fusion server & better error reporting

--HG--
branch : odoo8

brmasson committed:
removed "required" constraint;

--HG--
branch : odoo8

anael-lorimier committed:
2.0b1

--HG--
branch : odoo8

anael-lorimier committed:
Added tag TAG_2.0b1 for changeset 801e4bf837a2

--HG--
branch : odoo8

anael-lorimier committed:
Use the latest features of py3o.template with odoo8

--HG--
branch : odoo8

anael-lorimier committed:
TAG_2.0b2

--HG--
branch : odoo8

anael-lorimier committed:
Added tag TAG_2.0b2 for changeset 2e68df0df26d

--HG--
branch : odoo8

anael-lorimier committed:
i18n

--HG--
branch : odoo8

anael-lorimier committed:
2.0

--HG--
branch : odoo8

anael-lorimier committed:
Added tag TAG_8.0.2.0 for changeset 77b8df614816

--HG--
branch : odoo8

szekawong committed:
8.0.2.1
Fixed bug -> removed raise exception when no xml id found.

--HG--
branch : odoo8

szekawong committed:
Étiquette TAG_8.0.2.1 ajoutée à la révision 65b3c62a20f1

--HG--
branch : odoo8

gavrelj committed:
Restore direct rendering for native ODT output formats

--HG--
branch : odoo8

gavrelj committed:
Add search view for py3o_template

--HG--
branch : odoo8

gavrelj committed:
Only use direct rendering as a fallback if no py3o server is defined

--HG--
branch : odoo8

gavrelj committed:
Can set Py3o servers as inactive

--HG--
branch : odoo8

sbidoul committed:
Fix raise error

--HG--
branch : odoo8

gavrelj committed:
Fix py3o_template_fallback help string

--HG--
branch : odoo8

gavrelj committed:
Fix py3o_template_fallback help string

--HG--
branch : odoo8

szekawong committed:
Add oe_json_serializer in external_dependencies.
Fix get_user_variables call.

--HG--
branch : odoo8

szekawong committed:
Replace get_user_variables (=deprecated) with get_all_user_python_expression.
Add call on template render function.

--HG--
branch : odoo8

brmasson committed:
Remove dependency to oe_json_serializer due to packaging issues

--HG--
branch : odoo8

lmignon committed:
Move to report_py3o after import from HG

lmignon committed:
Merge branch 'report_py3o' into HEAD

lmignon committed:
remove hg files

lmignon committed:
Add setup

lmignon committed:
[FIX] View inheritance may not use attribute 'string' as a selector.

lmignon committed:
[IMP] New API + Short Hearder

lmignon committed:
[FIX] Add depndency on py3o.template

faide committed:
Merge pull request #1 from acsone/9.0-import_report_py3o

9.0 import report py3o

lmignon committed:
Remove unicode string

lmignon committed:
[IMP] Add the possiblity to get the template from an absolute path on the server

faide committed:
Merge pull request #2 from acsone/9.0-import_report_py3o

9.0 import report py3o

lmignon committed:
[IMP] Allow user to use local fusion

faide committed:
Merge pull request #3 from acsone/9.0-import_report_py3o

[IMP] Allow user to use local fusion

lmignon committed:
Add dependency on py3.formats for travis

faide committed:
Merge pull request #4 from acsone/9.0-import_report_py3o

Add dependency on py3.formats for travis

lmignon committed:
Add a simple way to extend the parser context

lmignon committed:
Remove ACL on removed model

lmignon committed:
Improve filetype selection name

lmignon committed:
Fix native flag

faide committed:
Merge pull request #5 from acsone/9.0-import_report_py3o

9.0 import report py3o

elicoidal commented on report_py3o/LICENSE:
```diff
@@ -0,0 +1,661 @@
+                    GNU AFFERO GENERAL PUBLIC LICENSE
```
@faide   Thanks for your contribution to the OCA. 
Let me introduce to you some basic information about contribution in OCA and I will review other points in more comments inside the code..
In general, please follow the principles: 
- Contribution guidelines: https://github.com/OCA/maintainer-tools/blob/master/CONTRIBUTING.md
- Please follow the OCA README template:
  https://github.com/OCA/maintainer-tools/blob/master/template/module/README.rst
- We recommend to include a screenshot as well to help new comers (nice to have)
- Please use simple headers in all py files following OCA templates
  https://github.com/OCA/maintainer-tools/blob/master/template/module/__openerp__.py#L3

In particular here, this file should be removed (normally it is at repo level)

elicoidal commented on report_py3o/README.md:
```diff
@@ -0,0 +1,18 @@
+### README ###
```
this should be transformed into README.rst template

elicoidal commented on report_py3o/__openerp__.py:
```diff
@@ -0,0 +1,34 @@
+# License AGPL-3.0 or later (http://www.gnu.org/licenses/agpl).
+{
+    'name': 'LibreOffice Report Engine',
+    'description': '''
```
remove =>README

elicoidal commented on report_py3o/security/ir.model.access.csv:
```diff
@@ -0,0 +1,5 @@
+access_py3o_template_admin,access_py3o_template_admin,model_py3o_template,base.group_no_one,1,1,1,1
+access_py3o_template_user,access_py3o_template_user,model_py3o_template,base.group_user,1,0,0,0
+access_py3o_server_admin,access_py3o_server_admin,model_py3o_server,base.group_no_one,1,1,1,1
+access_py3o_server_user,access_py3o_server_user,model_py3o_server,base.group_user,1,0,0,0
```
should not you add a blank line?

elicoidal reviewed (changes requested):
Thanks for the contribution.
some formal modification to fit the current OCA branding

pedrobaeza commented:
Can you please squash some commits together that are not meaningful in the final repository?

faide committed:
fixed missing comma in model definition

NL66278 commented:
Just some general comments. Researching what to do to get py3o uno to run I see a lot of confusing and incomplete information, examples not working, docker images that are no longer there etc. It makes me doubt whether this will really be an improvement compared to aeroo reports.

lmignon committed:
Add simple tests and fix call to fusion server

faide committed:
Merge pull request #7 from acsone/9.0-import_report_py3o

Add simple tests and fix call to fusion server

lmignon committed:
Fix skip fusion on server when the fusion is local

NL66278 commented:
```
Summary of the commands to install the server on debian jessie (run as root):
# aptitude install docker.io
# pip install docker-compose
# wget https://bitbucket.org/xcgd/py3o.fusion/raw/76dcb04477c4ddcaa303db94ae340561c2da75a9/docker-compose.yml
# export COMPOSE_API_VERSION=1.18
# docker-compose up
```

lmignon committed:
Improve tests

lmignon committed:
Replace openerp element by odoo

alexis-via commented:
@NL66278 I don't agree with you. The documentation, information, code quality, test suite of py3o is much better than the one of aeroo. Aeroo report doesn't really have any documentation at all, no unit tests, etc... Of course, as py3o is new for the vast majority of Odoo community members, those who plan to switch from aeroo to py3o will have to go through the learning curve, but it should be easier than aeroo for newcomers. If you came at the OCA code sprint and worked with the libreoffice-reporting group, you would have learnt many useful things, ask your questions to the py3o experts and participate in the choice of the OCA libreoffice-based reporting engine. But criticizing today on this PR when the work is still in progress is certainly not a good way to welcome a new contribution and thank the hard work of the authors of py3o.

faide committed:
Merge pull request #8 from acsone/9.0-import_report_py3o

9.0 import report py3o

alexis-via committed:
Small usability improvements

Replace README.md by README.rst (not finished)

alexis-via commented:
I just succedded in generating my first py3o PDF document with odoo9 + report_py3o of this PR + py3o.fusion server + py3o.render + libreoffice on my Ubuntu 16.04... I'm so happy ! :) Big thanks to @faide for his help to make it work. I'll document the steps and tricks to configure all that (but after my presentation tomorrow on e-invoice and e-orders). :+1: once my PR will be merged in faide's branch https://github.com/faide/reporting-engine/pull/9

lasley commented on report_py3o/README.md:
```diff
@@ -0,0 +1,18 @@
+### LibreOffice Report Engine for Odoo/OpenERP ###
+
+
+This module allows you to convert any type of report into LibreOffice/OpenOffice.
```
IMO LibreOffice/OpenOffice is too application specific. This module seems to convert into the [Open Document Format (ODF)](https://en.wikipedia.org/wiki/OpenDocument)

lasley commented on report_py3o/NEWS:
```diff
@@ -0,0 +1,4 @@
+report_py3o 1.3
```
Please remove this file

lasley commented on report_py3o/__openerp__.py:
```diff
@@ -0,0 +1,34 @@
+# Copyright 2013 XCG Consulting (http://odoo.consulting)
+# License AGPL-3.0 or later (http://www.gnu.org/licenses/agpl).
+{
+    'name': 'LibreOffice Report Engine',
```
Same note regarding the application specific name. This is converting into the Open Document Format, as there is no such thing as LibreOffice format. These same templates should also work just fine in Microsoft Office.

lasley commented on report_py3o/__openerp__.py:
```diff
@@ -0,0 +1,34 @@
+''',
+    'version': '9.0.1.0.0',
+    'category': 'Reporting',
+    'author': 'XCG Consulting',
```
Add `Odoo Community Association (OCA)`

lasley commented on report_py3o/__openerp__.py:
```diff
@@ -0,0 +1,34 @@
+    'author': 'XCG Consulting',
+    'website': 'http://odoo.consulting/',
+    'depends': [
+        'base',
```
base is implicit, please remove

lasley commented on report_py3o/models/ir_report.py:
```diff
@@ -0,0 +1,119 @@
+# -*- coding: utf-8 -*-
```
The name of this file should be the snake_case_version of the model name

lmignon replied:
@lasley IMO in this case the model name is too long... ir_action_report_xml.py

lasley commented on report_py3o/models/ir_report.py:
```diff
@@ -0,0 +1,119 @@
+from ..py3o_parser import Py3oParser
+
+
+class ReportXml(models.Model):
```
The name of this class should be the UpperCamelCased version of the model name

lasley commented on report_py3o/models/ir_report.py:
```diff
@@ -0,0 +1,119 @@
+
+    _inherit = 'ir.actions.report.xml'
+
+    @api.one
```
api.one is deprecated

alexis-via replied:
No, api.one is not deprecated any more.

lasley commented on report_py3o/models/ir_report.py:
```diff
@@ -0,0 +1,119 @@
+            raise ValidationError(
+                "Field 'Output Format' is required for Py3O report")
+
+    @api.one
```
api.one is deprecated

lasley commented on report_py3o/models/ir_report.py:
```diff
@@ -0,0 +1,119 @@
+    def _check_py3o_filetype(self):
+        if self.report_type == "py3o" and not self.py3o_filetype:
+            raise ValidationError(
+                "Field 'Output Format' is required for Py3O report")
```
Please add translation  to errors

lasley commented on report_py3o/models/ir_report.py:
```diff
@@ -0,0 +1,119 @@
+        if ((not is_native or not self.py3o_is_local_fusion) and
+                not self.py3o_server_id):
+            raise ValidationError(
+                "Can not use not native format in local fusion. "
```
Add translation

lasley commented on report_py3o/models/ir_report.py:
```diff
@@ -0,0 +1,119 @@
+# Copyright 2013 XCG Consulting (http://odoo.consulting)
+# License AGPL-3.0 or later (http://www.gnu.org/licenses/agpl).
+import os
+from py3o.formats import Formats
```
All external dependency imports should be guarded.

```
try:
    from py3o.formats import Formats
except ImportError:
    _logger.info('Py3o is not installed')
```

bealdav replied:
guidelines seems says .debug instead of .info

lasley commented on report_py3o/models/ir_report.py:
```diff
@@ -0,0 +1,119 @@
+            selections.append((name, description))
+        return selections
+
+    py3o_filetype = fields.Selection(
```
Field definitions should be at top of class

lasley commented on report_py3o/models/ir_report.py:
```diff
@@ -0,0 +1,119 @@
+            if not isinstance(new_report, Py3oParser):
+                new_report = None
+        else:
+            cr.execute(
```
This is possible with ORM. Was there a specific reason that SQL was used?

lasley commented on report_py3o/models/py3o_server.py:
```diff
@@ -0,0 +1,11 @@
+    _name = 'py3o.server'
+
+    url = fields.Char("URL", required=True)
+    is_active = fields.Boolean("Active", default=True)
```
Odoo convention for this field name is `active`, which will provide some special handling such as auto-hiding when deactivated

lasley commented on report_py3o/models/py3o_template.py:
```diff
@@ -0,0 +1,19 @@
+    _name = 'py3o.template'
+
+    name = fields.Char(required=True)
+    py3o_template_data = fields.Binary("LibreOffice template")
```
Same note re application specific text. This will work just fine in MS Office. Prefer the document format instead of the application name

lmignon replied:
@lasley py3o_template refers to the file used as template... IMO the name appropriate

lasley commented on report_py3o/models/py3o_server.py:
```diff
@@ -0,0 +1,11 @@
+from openerp import fields, models
+
+
+class Py3oServer(models.Model):
```
IMO we're switching naming schemas all over the place, and not many people will know what `py3o` is. Please switch this to ODFServer or similar

alexis-via replied:
No, we won't change model names.

lasley commented on report_py3o/models/py3o_template.py:
```diff
@@ -0,0 +1,19 @@
+from openerp import fields, models
+
+
+class Py3oTemplate(models.Model):
```
IMO we're switching naming schemas all over the place, and not many people will know what py3o is. Please switch this to ODFTemplate or similar

alexis-via replied:
No, we keep py3o. The py3o project has existed for many years, it's name won't be changed.

lasley commented on report_py3o/py3o_parser.py:
```diff
@@ -0,0 +1,217 @@
+from base64 import b64decode
+import requests
+from tempfile import NamedTemporaryFile
+from py3o.template.helpers import Py3oConvertor
```
Guard all external imports

lasley commented on report_py3o/py3o_parser.py:
```diff
@@ -0,0 +1,217 @@
+        """private helper to fetch the template data either from the database
+        or from the default template file provided by the implementer.
+
+        ATM this method takes a report definition recordset
```
I would spell out `At the moment` for clarity

lasley commented on report_py3o/py3o_parser.py:
```diff
@@ -0,0 +1,217 @@
+        to try and fetch the report template from database. If not found it
+        will fallback to the template file referenced in the report definition.
+
+        @param report_obj: a recordset representing the report defintion
```
This is mixing Doxygen and sphinx param format. Prefer sphinx:

```
:param report_obj:
```

Also definition is spelled wrong

lasley commented on report_py3o/py3o_parser.py:
```diff
@@ -0,0 +1,217 @@
+            for fct in _extender_functions[xml_id]:
+                fct(report_xml, parser_instance.localcontext)
+
+    def create_single_pdf(self, cr, uid, ids, data, report_xml, context=None):
```
Please use new api

lmignon replied:
@lasley here we are in the parser not in an odoo model. Therefore the new api is applicable and moreover our class extends a pure python class and we override only some methods... so we must keep the same method signature

lasley commented on report_py3o/py3o_parser.py:
```diff
@@ -0,0 +1,217 @@
+            r = requests.post(
+                report_xml.py3o_server_id.url, data=fields, files=files)
+            if r.status_code != 200:
+                # server says we have an issue... let's tell that to enduser
```
This comment does not add anything to the meaning of the code, the raising of an error implicitly says you are telling the user about an error. No need to repeat yourself IMO

lasley commented on report_py3o/py3o_parser.py:
```diff
@@ -0,0 +1,217 @@
+
+        return res, "." + filetype
+
+    def create(self, cr, uid, ids, data, context=None):
```
USe new api

lasley commented on report_py3o/views/ir_report.xml:
```diff
@@ -0,0 +1,31 @@
+<odoo>
+    <data>
```
You can remove this data tag

lasley commented on report_py3o/views/menu.xml:
```diff
@@ -0,0 +1,8 @@
+<?xml version="1.0" encoding="utf-8"?>
+<odoo>
+    <data>
```
Remove data tag

lasley commented on report_py3o/views/py3o_server.xml:
```diff
@@ -0,0 +1,42 @@
+<?xml version="1.0" encoding="utf-8"?>
+<odoo>
+    <data>
```
Remove data tag

lasley commented on report_py3o/views/py3o_template.xml:
```diff
@@ -0,0 +1,54 @@
+<?xml version="1.0" encoding="utf-8"?>
+<odoo>
+    <data>
```
Remove data tag

lasley reviewed (changes requested):
Thanks for the submission @faide - comments inline

Additionally and IMO, this module should be named `report_odf` - nobody is going to know what py3o is

rvalyi commented:
guys, I'm not sure where it is the best to comment on this but in the comparison doc you said Aeroo has no Docker image. Well At Akretion we are using https://github.com/akretion/aeroocker in several projects for nearly a year now. That being said we are open to migrate to py3o in the future if it proves better.

alexis-via committed:
Take into accounts most remarks of @lasley

Remove <data> in views
Protect import of py3o libs
Remove dep on base module
Other small changes

alexis-via commented:
I took into account most remarks of @lasley in my PR https://github.com/faide/reporting-engine/pull/9 against faide's branch.

faide committed:
Merge pull request #9 from akretion/py3o-usability

Small usability improvements

NL66278 commented:
@alexis-via I am sorry that the doubts I raised are seen as just criticizing the hard work of people I certainly respect. It is nor was my intention just to criticise from the sidelines. But then the point  about having reviews is also to raise issues when you encounter them. As you should know I did come to the codesprint and talk to the py3o people, that I found very helpful. I made two bug reports on the py3o project on bitbucket complete with documentation on what I did for the sake of validation. And I wrote a later comment on how I got finally the py3o fusion server to work on Debian.  Later on I created a buildout recipe to create an Odoo instance with py3o reporting enabled. It is here: https://github.com/nl66278/reporting-engine/tree/buildout-9.0-testing. That said I think the documentation still leaves much to be desired. I am fully willing to participate in rectifying that. Just the raw work of what I found out yesterday I will attach to this comment. Actually I wrote up a lot more, but the parts were I was led astray by the existing documentation I deleted.

NL66278 commented:
WTF: github does not support ODF, while supporting proprietary formats like docx.
[2016-10-04-odoo-p3o-reporting.pdf](https://github.com/OCA/reporting-engine/files/511274/2016-10-04-odoo-p3o-reporting.pdf)

Well: is has difficulty with pdf, but here is my writeup of what to do to get the py3o fusion server working om debian.

faide commented:
Hi Ronald I know documentation is still terse on the deployment parts but there are visible efforts and openness. Did you see the library usage documentation ?
http://py3otemplate.readthedocs.io
This is what we intend to have for the whole project.

I'll look into your issues and I am grateful we were able to discuss on them, I'll make sure to pin them down!

NL66278 commented:
Hello @faide yes I looked at that documentation that is certainly very useful. The part I was having an issue with was with the setup / installation part. If in any way my doubts seemed to apply a lack of appreciation for the great work done,  I would certainly apologize for that.

alexis-via commented:
@NL66278 Thanks for your answer. Glad to see that we are now all working on improving py3o to have the best code and a good doc.

lmignon committed:
Replace plain SQL by orm

lmignon committed:
rename module ir_report to ir_actions_report_xml

lmignon committed:
Add more tests and fixes issues found by tests

faide committed:
Merge pull request #10 from acsone/9.0-import_report_py3o

More tests...

lmignon commented:
@faide Can you provide a CLA for the following commiters?  http://odoo-community.org/page/website.cla

```
Vincent Lhote-Hatakeyama (no github login found)
@anael-lorimier (login unknown in OCA database)
@brmasson (login unknown in OCA database)
@faide (login unknown in OCA database)
@alexbrun (login unknown in OCA database)
@szekawong (login unknown in OCA database)
```

lasley commented:
Rebuilding Runbot now that dependencies have been installed

chermed commented:
Can someone explain to us : what is the difference between `Aeroo` And `py3o` ?

lmignon commented:
@chermed Basically the 2 modules provides the same functionalities... (see https://docs.google.com/spreadsheets/d/1YgniNHjEWDHXZkhYWwasxmIklW734TL143euQ2Sjepc/edit#gid= for details)
Nevertheless thanks to XCG report_py3o is proposed to be included under the OCA umbrella. Besides being a clean and well-tested module, the inclusion under the OCA umbrella ensures that the module will continue to evolve across the Odoo versions and the community will have the opportunity to contribute to its development.

lmignon committed:
[FIX] Check constrains only if report_type == 'py3o'

faide committed:
Merge pull request #11 from acsone/9.0-import_report_py3o

[FIX] Check constrains only if report_type == 'py3o'

alexis-via committed:
report_py3o: full installation procedure in README

alexis-via commented:
I made a PR with the installation procedure in README https://github.com/faide/reporting-engine/pull/12

faide committed:
Merge pull request #12 from akretion/9-report_py3o-README-updated

9.0 report_py3o: full installation procedure in README

alexis-via commented:
For me, this PR should be merged now (if the CLAs have been signed).

alexis-via commented:
@faide During the code sprint, we managed to call the display_address() method in py3o with

```
py3o://function="display_address(o.partner_id)"
```

but the carriage returns were not taken into account in the ODT file. For example, the current result is:

```
35B rue Montgolfier 69100 VILLEURBANNE France
```

with everything on the same line. Do you have an idea how we could fix that ? It makes the method display_address almost unusable in py3o for the moment...

alexis-via committed:
Add command line for libreoffice in README

faide committed:
Merge pull request #13 from akretion/9report_py3o-README-fix-missing

Add command line for libreoffice in README

eLBati commented:
Thanks for this work.

I am testing it and wondering how to avoid an empty field to be printed as **False**
How can I print an empty string?

lasley commented:
Changes LGTM. 

Once @elicoidal confirms good as well, can you please squash the logical commits - specifically making sure to change the commit message on the one that tags me (2fbbdc9). Commit messages that tag users will spam them indefinitely 😦

eLBati commented:
Thanks  @alexis-via  for the answer  https://twitter.com/alexisdlattre/status/791750379377754113

For now, we can anyway use the `if` py3o syntax
:+1:

JosDeGraeve commented:
Hi,

I have tested the p3o reporting engine and git it up and running in practically no time.  Good work !

For reports there is a flag "is_local_fusion".  I'm a bit confused by this.  In order to generate a PDF report, an external py3o server is needed for 'fusion/renderserver/oooserver'.  However, with an external fusion, it is not possible to call odoo methods: that is a big restriction.

Why is it not possible to have the fusion part done on the odoo side + have the conversion step to PDF or other formats done by the rendering server ?

A feature that aeroo reports had, was to generate multiple reports separately.  This allowed, for example, to print multiple invoices in a single run, but with each invoice generated in a separate document.  If you need per-document page numbering, or want to eliminate nasty page alignment issues between documents, this feature comes very handy.  Would it be hard to implement such a feature ? 

Last question is how are translations handled ?  I can't find any info on that.

Tnx

lmignon commented:
@josDeGraeve the flag 'is_local_fusion' is to use if you want to be able to call methods from your templates. The rendition and the fusion are 2 separates steps. Even if you select the local fusion you can specify a py3o server to render your template into pdf or doc.

scigghia commented:
hi, i've read documentation and source, but i need an example report module to study 
is there?

JosDeGraeve commented:
@scigghia for documentation, take a look at http://py3otemplate.readthedocs.io/en/latest/
As small sample doc can be found in  `demo/res_user.odt`

lmignon commented:
@scigghia @JosDeGraeve You can also take a look at https://bitbucket.org/faide/py3o.template/src/1e5efa93e8e64a642174c129c63f3735c9be3e6a/example/?at=default

lmignon commented:
A refactoring is proposed https://github.com/OCA/reporting-engine/pull/78#issuecomment-267570835
